### ISSUE TITLE
recompile with -fPIC when adding library dependent code as build-depenency of library

### ISSUE THREAD

**Issue by bneijt** [commercialhaskell/stack#1987]
My project compiles locally, however I can't get it to compile on Travis-CI. I have tried different versions of both stack and ghc, but can't seem to get a reproducable build.

Steps to reproduce:
- Use stack new to create a project
- Add a library as an external dependency of the "library" package in cabal
- Add the library dependency to stack.yml
- Add travis file
- Build locally and let Travis CI build it using the stack on Travis manual.

I have created an example repository with code to show the result:
https://github.com/bneijt/fpic
https://travis-ci.org/bneijt/fpic

Expected result:
    Normal compilation, no errors

Actual result
[1 of 1] Compiling Lib              ( src/Lib.hs, .stack-work/dist/x86_64-linux/Cabal-1.22.5.0/build/Lib.o )
Progress: 1/3/usr/bin/ld: .stack-work/dist/x86_64-linux/Cabal-1.22.5.0/build/Lib.dyn_o: relocation R_X86_64_PC32 against symbol `fpiczu2fjQ7h3pNG41YDOOD1F9B2_Lib_someFunc2_closure' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Bad value
collect2: ld returned 1 exit status
Progress: 2/3
--  While building package fpic-0.1.0.0 using:
      /home/travis/.stack/setup-exe-cache/x86_64-linux/setup-Simple-Cabal-1.22.5.0-ghc-7.10.3 --builddir=.stack-work/dist/x86_64-linux/Cabal-1.22.5.0 build lib:fpic exe:fpic-exe test:fpic-test --ghc-options " -ddump-hi -ddump-to-file"

    Process exited with code: ExitFailure 1

The FAQ mentions this problem, but the problem is reproducible and seems to be related to using a package that requires an external library as a dependency.

**Comment by bneijt** [comment]
Installing with cabal works in a local Travis-CI docker image:

```
docker run -it quay.io/travisci/travis-ruby /bin/bash
su - travis
sudo apt-get install libgmp-dev c2hs libssh2-1-dev cabal-install
cabal update
```

change the fpic.cabal file to allow for base 4.5

```
cabal install
```
[cabal install verbose.txt](https://github.com/commercialhaskell/stack/files/201350/cabal.install.verbose.txt)

Stack install still fails with the same error:
[stack install verbose.txt](https://github.com/commercialhaskell/stack/files/201351/stack.install.verbose.txt)

**Comment by mgsloan** [comment]
Thanks for the repo!  Does `--ghc-options -fPIC` fix the issue?

**Comment by bneijt** [comment]
No, starting with `stack build --verbose --test --ghc-options -fPIC` (line 368 in Travis build 3) fails, but outputs:
```
--  While building package fpic-0.1.0.0 using:

      /home/travis/.stack/setup-exe-cache/x86_64-linux/setup-Simple-Cabal-1.22.5.0-ghc-7.10.3 --builddir=.stack-work/dist/x86_64-linux/Cabal-1.22.5.0 build lib:fpic exe:fpic-exe test:fpic-test --ghc-options " -ddump-hi -ddump-to-file"

    Process exited with code: ExitFailure 1
```
so it seems the ghc-options did not trickle down to the failing command.

**Comment by mgsloan** [comment]
ghc-options get provided to the configure step, not the build step.

**Comment by bneijt** [comment]
After reading
http://stackoverflow.com/questions/19768267/relocation-r-x86-64-32s-against-linking-error
and looking into libssh2-1-dev package used by Travis CI it seems that the problem might just be the libssh2 package not having the right binaries.

**Comment by mgsloan** [comment]
Sounds likely!

**Comment by bneijt** [comment]
I'm still baffled, but I did find out the following: moving the libssh2 dependency from the "library" to the "executable" fixes it (on Travis CI), however I need to use the libssh2 dependency in the library, not the executable.

**Comment by bneijt** [comment]
Just confirmed that installing a later version of libssh2 (from source) fixes the problem: libssh2-1.4.3 from source works 1.2.8-2 from APT package repository does not.
